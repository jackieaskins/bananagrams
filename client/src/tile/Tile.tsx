import { TILE_SIZE } from './constants';

const Tile = ({ letter }: { letter: string }): JSX.Element => (
  <div
    className="no-drag"
    css={{
      width: `${TILE_SIZE}px`,
      height: `${TILE_SIZE}px`,
      backgroundColor: '#ffffc7',
      border: '0.5px solid black',
      borderRadius: '5px',
      flexShrink: 0,
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
    }}
  >
    {letter}
  </div>
);

export default Tile;
