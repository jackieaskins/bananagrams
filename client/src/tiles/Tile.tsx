import { Typography, Box } from '@mui/material';
import { useDrag } from 'react-dnd';

import { Tile as TileType } from './types';

type TileProps = {
  boardLocation: { x: number; y: number } | null;
  tile: TileType;
  color?: string;
};

const Tile: React.FC<TileProps> = ({
  boardLocation,
  tile: { id, letter },
  color = 'black',
}) => {
  const [{ isDragging }, dragRef] = useDrag({
    item: { type: 'TILE', id, boardLocation },
    collect: (monitor) => ({
      isDragging: monitor.isDragging(),
    }),
  });

  const margin = !!boardLocation ? '0' : '5px';

  return (
    <Box
      ref={dragRef}
      display="inline-flex"
      justifyContent="center"
      alignItems="center"
      border={1}
      borderRadius={1}
      sx={{
        color: color,
        backgroundColor: '#ffffc7',
        cursor: 'move',
        height: '25px',
        margin,
        opacity: isDragging ? 0.5 : 1,
        width: '25px',
      }}
    >
      <Typography variant="body2">{letter}</Typography>
    </Box>
  );
};

export default Tile;
