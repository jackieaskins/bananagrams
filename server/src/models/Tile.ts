import BaseModel from './BaseModel';

export type TileJSON = {
  id: string;
  letter: string;
};

export default class Tile implements BaseModel<TileJSON> {
  private id: string;
  private letter: string;

  constructor(id: string, letter: string) {
    this.id = id;
    this.letter = letter;
  }

  getId(): string {
    return this.id;
  }

  getLetter(): string {
    return this.letter;
  }

  toJSON(): TileJSON {
    const { id, letter } = this;
    return { id, letter };
  }

  reset(): void {
    return;
  }
}
