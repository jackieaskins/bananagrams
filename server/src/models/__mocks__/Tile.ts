export default jest.fn().mockImplementation((id, letter) => ({
  getId: jest.fn().mockReturnValue(id),
  getLetter: jest.fn().mockReturnValue(letter),
  toJSON: jest.fn().mockReturnValue(`TileJSON - ${id} ${letter}`),
}));
