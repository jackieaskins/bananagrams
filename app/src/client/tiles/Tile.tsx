import { Box, Text } from "@chakra-ui/react";
import { useDrag } from "react-dnd";
import { Tile as TileType } from "@/types/tile";

type TileProps = {
  boardLocation: { x: number; y: number } | null;
  tile: TileType;
  color?: string;
};

export default function Tile({
  boardLocation,
  tile: { id, letter },
  color = "black",
}: TileProps): JSX.Element {
  const [{ isDragging }, dragRef] = useDrag({
    item: { type: "TILE", id, boardLocation },
    collect: (monitor) => ({
      isDragging: monitor.isDragging(),
    }),
  });

  const margin = boardLocation ? "0" : "5px";

  return (
    <Box
      ref={dragRef}
      alignItems="center"
      backgroundColor="yellow.100"
      border="1px"
      borderColor={color}
      borderRadius="5px"
      color={color}
      cursor="move"
      display="inline-flex"
      height="25px"
      justifyContent="center"
      margin={margin}
      opacity={isDragging ? 0.5 : 1}
      width="25px"
    >
      <Text as="b">{letter}</Text>
    </Box>
  );
}
