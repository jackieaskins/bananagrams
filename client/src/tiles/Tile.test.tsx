import { shallow, ShallowWrapper } from "enzyme";
import { useDrag } from "react-dnd";
import Tile from "./Tile";

jest.mock("react-dnd", () => ({
  useDrag: jest.fn().mockReturnValue([{ isDragging: false }, jest.fn()]),
}));

const DEFAULT_BOARD_LOCATION = { x: 0, y: 0 };
const DEFAULT_ID = "A1";

describe("<Tile />", () => {
  const renderComponent = (propOverrides = {}): ShallowWrapper =>
    shallow(
      <Tile
        boardLocation={DEFAULT_BOARD_LOCATION}
        tile={{ id: DEFAULT_ID, letter: "A" }}
        {...propOverrides}
      />,
    );

  test("renders properly", () => {
    expect(renderComponent()).toMatchSnapshot();
  });

  describe("useDrag", () => {
    beforeEach(() => {
      renderComponent();
    });

    test("is called with correct params", () => {
      expect(useDrag).toHaveBeenCalledWith({
        item: {
          id: DEFAULT_ID,
          boardLocation: DEFAULT_BOARD_LOCATION,
          type: "TILE",
        },
        collect: expect.any(Function),
      });
    });

    test("collect retrieves isDragging from monitor", () => {
      const monitor = {
        isDragging: jest.fn().mockReturnValue("isDragging"),
      };

      expect(useDrag.mock.calls[0][0]["collect"](monitor)).toEqual({
        isDragging: "isDragging",
      });
    });
  });

  describe("margin", () => {
    test("is set to 0 when on board", () => {
      expect(renderComponent().props().margin).toBe("0");
    });

    test("is set when not on board", () => {
      expect(renderComponent({ boardLocation: null }).props().margin).toBe(
        "5px",
      );
    });
  });

  describe("opacity", () => {
    test("is set to 1 when not dragging", () => {
      expect(renderComponent().props().opacity).toBe(1);
    });

    test("is dimmed while dragging", () => {
      useDrag.mockReturnValue([{ isDragging: true }, jest.fn()]);
      expect(renderComponent().props().opacity).toBe(0.5);
    });
  });

  describe("color", () => {
    test("is set to black by default", () => {
      expect(renderComponent().props().color).toBe("black");
    });

    test("is set to passed in color", () => {
      const color = "red";
      expect(renderComponent({ color }).props().color).toEqual(color);
    });
  });
});
