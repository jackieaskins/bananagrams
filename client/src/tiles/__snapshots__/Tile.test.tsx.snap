// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tile /> renders properly 1`] = `
<ForwardRef(Box)
  alignItems="center"
  border={1}
  borderRadius={1}
  display="inline-flex"
  justifyContent="center"
  sx={
    Object {
      "backgroundColor": "#ffffc7",
      "color": "black",
      "cursor": "move",
      "height": "25px",
      "margin": "0",
      "opacity": 1,
      "width": "25px",
    }
  }
>
  <ForwardRef(Typography)
    variant="body2"
  >
    A
  </ForwardRef(Typography)>
</ForwardRef(Box)>
`;
